docs/freqai.md

```diff
@@ -807,3 +807,85 @@ Code review, software architecture brainstorming:
 Beta testing and bug reporting:
 @bloodhunter4rc, Salah Lamkadem @ikonx, @ken11o2, @longyu, @paranoidandy, @smidelis, @smarm,
 Juha Nykänen @suikula, Wagner Costa @wagnercosta
+
+## Using the `spice_rack`
+
+<!-- Dont forget this section during the doc reorg! -->
+
+The `spice_rack` is aimed at users who do not wish to deal with setting up `FreqAI` confgs, but instead prefer to interact with `FreqAI` similar to a `talib` indicator. In this case, the user can instead simply add two keys to their config:
+
+```json
+    "freqai_spice_rack": true, 
+    "freqai_identifier": "spicey-id",
+```
+
+Which tells `FreqAI` to set up a pre-set `FreqAI` instance automatically under the hood with preset parameters. Now the user can access a suite of custom `FreqAI` supercharged indicators inside their strategy:
+
+```python
+        dataframe['dissimilarity_index'] = self.freqai.spice_rack(
+            'DI_values', dataframe, metadata, self)
+        dataframe['maxima'] = self.freqai.spice_rack(
+            '&s-maxima', dataframe, metadata, self)
+        dataframe['minima'] = self.freqai.spice_rack(
+            '&s-minima', dataframe, metadata, self)
+        self.freqai.close_spice_rack()  # user must close the spicerack
+```
+
+Users can then use these columns, concert with all their own additional indicators added to `populate_indicators` in their entry/exit criteria and strategy callback methods the same way as any typical indicator. For example:
+
+```python
+    def populate_entry_trend(self, df: DataFrame, metadata: dict) -> DataFrame:
+
+        df.loc[
+            (
+                (df['dissimilarity_index'] < 1) &
+                (df['minima'] > 0.1)
+            ),
+            'enter_long'] = 1
+
+        df.loc[
+            (
+                (df['dissimilarity_index'] < 1) &
+                (df['maxima'] > 0.1)
+            ),
+            'enter_short'] = 1
+
+        return df
+
+    def populate_exit_trend(self, df: DataFrame, metadata: dict) -> DataFrame:
+
+        df.loc[
+            (
+                (df['dissimilarity_index'] < 1) &
+                (df['maxima'] > 0.1) 
+            ),
+
+            'exit_long'] = 1
+
+        df.loc[
+            (
+
+                (df['dissimilarity_index'] < 1) &
+                (df['minima'] > 0.1)
+            ),
+            'exit_short'] = 1
+
+        return df
+```
+
+The user does need to ensure their `informative_pairs()` contains the following (users can add their own `informative_pair` needs to the bottom of this template):
+
+```python
+    def informative_pairs(self):
```

Comment:
i think the docs are missing a crucial piece.

Based on the docs - i can use any strategy (say - the default sample strategy), throw in the 2 config options below - and use the indicators.

That'll however mean that `self.freqai.start()` will never be called (nobody told me to add that to `populate_indicators()`.

as that's "default" code (every freqAI is expected to do the same) - it might be worth moving that to the "backend" - no longer requiring the user to call `.freqai.start()`.

The same could (probably should or must) go for informative-pairs - the moment we ask a user "just copy/paste this segment into your strategy" - it's something that'll become a problem later on.
Imagine we see a bug in this segment or the need to change something for whatever reason (this code is now copy/pasted in everyone's strategy - good luck migrating without a huge mess).

---

I think we should consider moving these 2 things to the backend (as a separate PR - which ideally is merged before this one) - while a breaking freqAI change - it'll be a huge interface simplification.
let me know what you think about that ...

Reply:
I don't understand, why are you saying that the spice_rack user needs to call `self.freqai.start()`? That's not how spice_rack interface is designed - see how spice_rack calls start in the backend...

Why are you suggesting normal FreqAI users shouldn't have to call freqai.start()? 

Both usecases are well defined and require their respective interactions with spice_rack() or start(). I don't see why we would hide either of these interactions from the user. Please elaborate. 

Re: informative pairs:

While I don't disagree with your point. Well we could do just as you say, and it is easy to add - and I considered it during this PR (after all it's the same code as I'm doing for `populate_any_indicators` 
). However, it means we take full control over `informative_pairs`, and as such, the spice_rack user can't add their own informative needs. Thus, imo it's best to allow them to add what ever they need to `inforative_pairs` since the idea is that spice_rack is just an another indicator.

Maybe an alternative solution is to add a `freqai_informative_pairs()` to `interface.py` and ensure it is always called before the typical `informative_pairs`?

Reply:
> However, it means we take full control over informative_pairs, 

that's clearly not the case if done properly (you shouldn't replace the function as you're doing with populate_any_indicators 
now).
the `@informative` decorator does something similar - so there's one place we can hook this into - and it's well defined

``` pseudo-code
if freqai: 
  # add these pair/timeframe combos unless they're already in the list
```


> Both usecases are well defined and require their respective interactions with spice_rack() or start().

Well i've so far not seen any usecase where `freqai.start()` is not at the top of `populate_indicators()` - so maybe the requirement/usecase for this isn't written anywhere I've looked so far.
based on my information, i therefore don't think that's a requirement (it's dummy code we require people to copy around) - but an interface decision.

Reply:
Re: decorator : 

Ok I will need to look into it and get back to you with a reply. Sounds like it is a good solution.

Re: location of FreqAI.start()

It can be put anywhere in populate_indicators(), there's nothing forcing it to be set at the top. 

So I'm still confused as to what you are asking for. are you saying we should consider removing these from populate indicators entirely?

Reply:
>  are you saying we should consider removing these from populate indicators entirely?

After a quick review here - my assumption was I'll need `freqai.start()` in indicators. that's not the case - which is fine.
I'm ... putting it up for discussion though. For now - it seems to be a "users must copy this to their strategy" solution - which is usually a solution we don't need.

> Ok I will need to look into it and get back to you with a reply. Sounds like it is a good solution.

don't worry - i can implement this (separate PR) as i know roughly where it needs to go

Reply:
>For now - it seems to be a "users must copy this to their strategy" solution - which is usually a solution we don't need.

Honestly - it is a fair point. We could automate everything from the `populate_indicators` side and the user just magically *gets* the predictions etc inside their dataframe, which they then append their own indicators to. 

My only concern is that it increases opacity which transitively increases the barrier to entry for new users. 

Typical FT user is someone who writes:

```
dataframe['rsi'] = ta.RSI(dataframe)
```

So i like that they can see "ok my dataframe goes into freqai, and comes out with new stuff attached

```
dataframe = self.freqai.start(dataframe)
```

(Where `spice_rack` aims to increase transparency even more)

Now we ask them to essentially "trust" that some stuff is gonna be in their dataframe. I am just unsure about if it will really improve the interface. But I want to hear your thoughts and responses to these points. May be worth asking the community directly - it is their code after all. 

I find the same issue with the `@informative` decorator. It kind "magically" makes stuff happen, where as a long time FT user I liked having control over where and when informatives were accessbile. This was due to three things:

1) harder to relearn a new method which requires reading up on
2) I now have to trust the authors of the decorator that i am getting the dataframe that I wanted.
3) the actual strategy code requires a bit of rethinking to accommodate the @informative decorator.

I therefore, never switched to using the @informative decorator. But, I am also not really using FT strats to intensely these days so my judgement isn't necessarily representative of the majority.

Reply:
`@informative` is a helper - simplifying the usecase (you no longer need to forget to put the informative pair in 2 locations (where you use it, and in the `informative_pairs()` location).

We didn't remove the old method though - so everyone is free to use whatever they like.
In this case it's however where we ask users to "do the same thing in every strategy" - which provides 0 benefit / 0 flexibility (if you forget this part - it'll fail with probably random errors).

I'll have a go at the indicators stuff - you can then let me know what you think (it shouldn't be a breaking change, either).

> (Where spice_rack aims to increase transparency even more)

for now - that's what i don't really see, but rather the opposite (hiding stuff from the beginner user is not a bad thing though).
For now, it's populating some magic columns for me - but the exact configs are unknown unless i look into the code.

Now that's a problem for another day - but we'll for sure have to have the exact config / setup in the docs - so i could go from a working spice-rack to a regular freqAI strategy once i'm "grown up" - and want to configure my own parameters.

> I now have to trust the authors of the decorator that i am getting the dataframe that I wanted.

this is exactly what spice-rack brings - users have to trust us / freqtrade / freqAI to give them the dataframe they never asked for.
but in the end, it's the same with the whole application - users have to trust the authors / reviewers to not break the application in unexpected ways, change behavior suddenly - or intentionally do malicious stuff. Without the trust - every user must read (and understand) the full sourcecode - while it's a possibility with open source - it's also not the reality of open source.

Reply:
> Now that's a problem for another >day - but we'll for sure have to have >the exact config / setup in the docs - >so i could go from a working spice->rack to a regular freqAI strategy once >i'm "grown up" - and want to >configure my own parameters.

Agreed 100%

> For now, it's populating some magic > columns for me - but the exact >configs are unknown unless i look >into the code.

> this is exactly what spice-rack >brings - users have to trust us / >freqtrade / freqAI to give them the >dataframe they never asked for.

I'll say this: spice rack lets users ask for an exact indicator and add that to their dataframe as a column. That is pure transparence. Sure it's black box generated data but at least the command lets them decide what they want. It isn't much different with some of those trading view API indicators. the goal is that they look for patterns themselves. As I explicitly point out in the doc - this is meant to be used as a supporting indicator only. 

This is geared toward the users who just want to get a generalized config and continue with their normal TA. Luckily we now have enough experience to set up a generalized config for them. 

I do, however, see spice rack evolving to let the user select from, say, 4 premade configs. I already leave that evolution open here with the config param "freqai_config" which defaults to lightgbm.

Reply:
> I'll have a go at the indicators stuff - you can then let me know what you think (it shouldn't be a breaking change, either).

Sounds good, I appreciate the help